### PR TITLE
DNS improvements (add compression & cleanup) + Fix DNSRRSRV + tests

### PR DESCRIPTION
**Pull request by gpotter2**
This PR is an attempt of adding DNS compression to scapy.

- Remove duplicated code
- Add DNS compression support: `dns_compress(pkt)` function
- Fix DNSRRSRV
- Rename DNSgetstr into dns_get_str (with deprecated old DNSgetstr)
- Adds `clear_cache` function, required to test DNS compression

### CODE REVIEW

**Comment by gpotter2** [comment]
@guedou Ready for review. The PR also reduces quite a lot of duplicates code

**Comment by gpotter2** [comment]
I managed to add full DNS compression support. Looking quite good.

I also tried to make the function simpler to read, using generators.

File: scapy/layers/dns.py [hunk 1]
```diff
@@ -19,12 +19,137 @@
...
+    if rburned:
+        return name, p, burned
+    return name, p
+
+
+def DNScompress(pkt):
+    """This function compresses a DNS packet according to its rules.
+
+    WARNING:
+    This function currently only compress strings if the hole string
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/layers/dns.py]: Tried to make it more explicit. Also added full compression support !

File: scapy/layers/dns.py [hunk 2]
```diff
@@ -19,12 +19,137 @@
...
 from scapy.modules.six.moves import range
 
 
+def DNSgetstr(s, p, pkt=None, rburned=False):
+    """This function decompresses a string s, from the character p.
+    params:
+     - s: the string to decompress
+     - p: start index of the string
+     - pkt: (optional) an InheritOriginDNSStrPacket packet
+     - rburned: (optional) return a third argument, the number of burned bytes
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/layers/dns.py]: That was an old commit

File: scapy/layers/dns.py [hunk 3]
```diff
@@ -19,12 +19,148 @@
...
 from scapy.layers.inet import IP, DestIPField, UDP, TCP
 from scapy.layers.inet6 import DestIP6Field
-from scapy.error import warning
+from scapy.error import warning, Scapy_Exception
 from functools import reduce
 import scapy.modules.six as six
 from scapy.modules.six.moves import range
 
 
+def DNSgetstr(s, p, pkt=None, _internal=False):
```
Comment (p-l-) [inline scapy/layers/dns.py]: Could you rename the `DNSxxx` functions to `dns_xxx`?

Comment (gpotter2) [reply scapy/layers/dns.py]: Of course. Let’s just hope that no one was using DNSgetstr (it wasn’t documented anyways).

I kept the original format

Comment (p-l-) [reply scapy/layers/dns.py]: I think you can safely rename `DNSgetstr`. If you want, you can add a `DNSgetstr` function that calls the new one and add a deprecation warning.

**Comment by gpotter2** [comment]
@guedou Please have a look at this

File: scapy/layers/dns.py [hunk 4]
```diff
@@ -19,12 +19,154 @@
...
 from scapy.layers.inet import IP, DestIPField, UDP, TCP
 from scapy.layers.inet6 import DestIP6Field
-from scapy.error import warning
+from scapy.error import warning, Scapy_Exception
 from functools import reduce
 import scapy.modules.six as six
 from scapy.modules.six.moves import range
 
 
+def dns_get_str(s, p, pkt=None, _internal=False):
```
Comment (guedou) [inline scapy/layers/dns.py]: Could you add some tests that will check the behavior of this function?

File: scapy/layers/dns.py [hunk 5]
```diff
@@ -19,12 +19,154 @@
...
+     - s: the string to decompress
+     - p: start index of the string
+     - pkt: (optional) an InheritOriginDNSStrPacket packet
+
+    returns: (decoded_string, end_index, left_string)
+    """
+    max_length = len(s)
+    name = b""
+    burned = 0
+    q = 0
```
Comment (guedou) [inline scapy/layers/dns.py]: Could you use longer/meaningful variable names?

File: scapy/layers/dns.py [hunk 6]
```diff
@@ -19,12 +19,154 @@
...
+    jpath = [p]
+    while True:
+        if p >= max_length:
+            warning("DNS RR prematured end (ofs=%i, len=%i)" % (p, len(s)))
+            break
+        l = orb(s[p])  # current value of the string at p
+        p += 1
+        burned += 1
+        if l & 0xc0:  # Label pointer
+            if not q:
```
Comment (guedou) [inline scapy/layers/dns.py]: Replace with `q == 0` ?

File: scapy/layers/dns.py [hunk 7]
```diff
@@ -19,12 +19,154 @@
...
+     - p: start index of the string
+     - pkt: (optional) an InheritOriginDNSStrPacket packet
+
+    returns: (decoded_string, end_index, left_string)
+    """
+    max_length = len(s)
+    name = b""
+    burned = 0
+    q = 0
+    jpath = [p]
```
Comment (guedou) [inline scapy/layers/dns.py]: Rename as "process_pointers" ?

File: scapy/layers/dns.py (hunk 4 above)
Comment (guedou) [reply scapy/layers/dns.py]: Could you also document the meaning of _internal ?

Comment (gpotter2) [reply scapy/layers/dns.py]: Added. Also changed a few var names

File: scapy/layers/dns.py [hunk 8]
```diff
@@ -19,12 +19,161 @@
...
+            if pkt and hasattr(pkt, "_orig_s") and pkt._orig_s:
+                # There should not be a loop as pkt is None
+                name += dns_get_str(pkt._orig_s, p, None, _internal=True)[0]
+                if burned == max_length:
+                    break
+            elif p in processed_pointers:
+                warning("DNS decompression loop detected")
+                break
+            elif not _internal:
+                raise Scapy_Exception("DNS message can't be compressed at this point!")  # noqa: E501
```
Comment (guedou) [inline scapy/layers/dns.py]: Could you rewrite this line to remove the noqa ?
For example:
msg = "DNS message can't be compressed at this point!"
raise Scapy_Exception(msg)

File: scapy/layers/dns.py [hunk 9]
```diff
@@ -19,12 +19,161 @@
...
+        # replacements is the list of all tuples (layer, field name)
+        # where this string was found
+        replace_pointer = replacements.pop(0)[2]
+        # replace_pointer is the packed pointer that should replace
+        # those strings. Note that pop remove it from the list
+        for rep in replacements:
+            # setfieldval edits the value of the field in the layer
+            val = rep[0].getfieldval(rep[1])
+            assert val.endswith(ck)
+            new_val = dummy_dns.i2m(None, val[:-len(ck)])[:-1] + \
```
Comment (guedou) [inline scapy/layers/dns.py]: I think that it is easier to read if written:
tmp = val[:-len(ck)])[:-1] + replace_pointer
new_val = dummy_dns.i2m(None, tmp)

**Comment by guedou** [comment]
I made two minors comments. When fixed, this PR can be merged.

File: scapy/layers/dns.py (hunk 9 above)
Comment (gpotter2) [reply scapy/layers/dns.py]: Hmm not exactly the same thing (you’re calling the i2m on the replace_pointer)

**Comment by guedou** [comment]
Thanks for this PR!

**Pull request merged**